chore(cc_rh_subscription): deprecate hyphenated fields, remove self.log

mostafaCamel opened the pull request:
## Proposed Commit Message
```
chore(cc_rh_subscription): deprecate hyphenated fields, remove self.log

Refs GH-6370
```
## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

blackboxsw commented on cloudinit/config/schemas/schema-cloud-config-v1.json:
```diff
@@ -666,6 +666,47 @@
+    },
+    "add_pool.properties": {
+      "type": "array",
+      "description": "A list of pools ids add to the subscription.",
```
```suggestion
      "description": "A list of pool IDs add to the subscription.",
```

mostafaCamel replied:
Done

blackboxsw reviewed (changes requested):
Thank you much @mostafaCamel for all the updates and example doc changes to make this happen. It looks really good. I think we need a few more adjustments to make this something we are ready to land.

mostafaCamel commented on cloudinit/config/schemas/schema-cloud-config-v1.json:
```diff
@@ -2604,48 +2656,182 @@
             "rhsm-baseurl": {
-              "type": "string",
-              "description": "Sets the baseurl in ``/etc/rhsm/rhsm.conf``."
+              "allOf": [
+                {
+                  "$ref": "#/$defs/rhsm_baseurl.properties"
+                },
+                {
+                  "deprecated": true,
+                  "deprecated_version": "25.3",
+                  "deprecated_description": "Use **rhsm_baseurl** instead."
+                }
+              ]
```
Fair enough. I was following the example of [ca_certs](https://github.com/canonical/cloud-init/blob/1b0cb0b851bc5b7805efeba8257f22f6064dee5f/cloudinit/config/schemas/schema-cloud-config-v1.json#L1232-L1241)

mostafaCamel replied:
After I flattened the structures, [the lowest-supported unit tests](https://github.com/canonical/cloud-init/actions/runs/17906568884/job/50908723014?pr=6470) began to fail
```
            with pytest.raises(SchemaValidationError, match=error_msg):
>               validate_cloudconfig_schema(config, get_schema(), strict=True)
E               Failed: DID NOT RAISE <class 'cloudinit.config.schema.SchemaValidationError'>
```

I looked at the ubuntu_advantage unit test and the unit test [is being skipped for jsonschmea<4.0](https://github.com/canonical/cloud-init/blob/1b0cb0b851bc5b7805efeba8257f22f6064dee5f/tests/unittests/config/test_cc_ubuntu_pro.py#L456) so I am doing similarly for the test in rh_subscription.

I am [splitting](https://github.com/canonical/cloud-init/pull/6470/files#diff-cef931f1d02d0cf72ce3a95aab8b292c4ec29e507ebffe85f46eebdf0983a54fR453) this parametrized test into a new test method as it's not possible (at least to my knowledge) to combine parametrized tests with custom skipIf decorators

mostafaCamel commented on cloudinit/config/cc_rh_subscription.py:
```diff
@@ -14,6 +14,19 @@
+    ("server-hostname", "server_hostname"),
+    ("auto-attach", "auto_attach"),
+    ("service-level", "service_level"),
+]
```
There is an issue with the method `_verify_keys()` as it checks valid_keys against self.rhel_cfg and not against the python object fields.

I thought about writing  new util method to replace all hyphens with underscores in self.rhel_cfg but I didn't go this route as I thought it's better to conserve self.rhel_cfg in the pristine state of the config that the user passed. Maybe I am wrong in pursuing this goal?

Anyways, this is what I did in my new commit:
- Did not change self.rhel_cfg but rather I am replacing hyphens with underscores [only within _verify_keys()](https://github.com/canonical/cloud-init/pull/6470/files#diff-a7df97d2c926815f722320b4f595b04aeb88663143b11219e2760229929f109aR149)
- Notice that the error message still prints the original key. I am also validating this behavior in [one of the existing unit tests](https://github.com/canonical/cloud-init/pull/6470/files#diff-cef931f1d02d0cf72ce3a95aab8b292c4ec29e507ebffe85f46eebdf0983a54fR311)
- Added [a new unit test](https://github.com/canonical/cloud-init/pull/6470/commits/7fa6ce6fdb97bec457cb7c36e4c92f2144e1b07d#diff-cef931f1d02d0cf72ce3a95aab8b292c4ec29e507ebffe85f46eebdf0983a54fR92) to confirm that the functionality does not change with deprecated keys (I probably should have written that test from the very beginning. My bad)

mostafaCamel commented:
Thanks @blackboxsw for the review. I addressed all your comments but I probably still need your validation on these 2 issues ([here](https://github.com/canonical/cloud-init/pull/6470#discussion_r2367760213) and [here](https://github.com/canonical/cloud-init/pull/6470#discussion_r2367563716))

mostafaCamel commented on cloudinit/config/cc_rh_subscription.py:
```diff
@@ -73,60 +86,61 @@ def handle(name: str, cfg: Config, cloud: Cloud, args: list) -> None:
+        ["org", "username", "password", "release_version"]
+        + [x[1] for x in DEPRECATION_TUPLES]
+        + [x[0] for x in DEPRECATION_TUPLES]
+    )
```
Done

mostafaCamel commented on cloudinit/config/schemas/schema-cloud-config-v1.json:
```diff
@@ -666,6 +666,47 @@
       },
       "minProperties": 1
     },
+    "activation_key.properties": {
```
Done

mostafaCamel commented on cloudinit/config/cc_rh_subscription.py:
```diff
@@ -135,30 +145,31 @@ def _verify_keys(self):
         for k in self.rhel_cfg:
-            if k not in self.valid_rh_keys:
+            k_with_no_hyphen = k.replace("-", "_")
+            if k_with_no_hyphen not in self.valid_rh_keys:
```
That's definitely the cleanest way to validate invalid key absence in this module.

I am thinking whether we can go a step further and remove this validatition altogether.

This [validation](https://github.com/canonical/cloud-init/blame/016abc02a4f1cb7c99f52744ebdbf16efb6f2c14/cloudinit/config/cc_rh_subscription.py#L137) was added 10 years ago when the file was first created (I presume at the time there was no schema-level additionalProperties validation) .

But then 3 years ago, James added  to the [rh_subscription section in the schema](https://github.com/canonical/cloud-init/blame/016abc02a4f1cb7c99f52744ebdbf16efb6f2c14/cloudinit/config/schemas/schema-cloud-config-v1.json#L2577-L2579) and added in the same commit a [unit test](https://github.com/canonical/cloud-init/blame/016abc02a4f1cb7c99f52744ebdbf16efb6f2c14/tests/unittests/config/test_cc_rh_subscription.py#L366-L367) to confirm the schema-level validation (these 2 links are permalinks obtained from the main branch)


So I guess we can remove this module-level validation as it is redundant and we already have upstream a schema-level validation doing the same thing. 

I pushed a [commit](https://github.com/canonical/cloud-init/pull/6470/commits/ce844a954c8d6bfce56a4af4cb10470ea56bdd13). Let me know what you think.
My bad, I should have thought about this in the initial commit instead of going the route of adding the deperecated keys to valid_rh_keys

blackboxsw replied:
Thanks for the improvement thoughts here @mostafaCamel. This dropping of invalid key validation represents a slight change in behavior in that former behavior would early-exit in error and not attempt to register the subscription if invalid keys were provided. But, I agree, we want cloud-init's cc_rh_subscription module to at least try to correctly register the subscription in the event that the required keys are present. I'm ok with cloud-init ignoring invalid keys during handle as that non-specific schema warning is loged anyway due to [cmd/main.py validate_cloudconfig_schema](https://github.com/canonical/cloud-init/blob/main/cloudinit/cmd/main.py#L663) as  `You may run 'sudo cloud-init schema --system' to check the details`[[1](https://github.com/canonical/cloud-init/blob/main/cloudinit/config/schema.py#L797)] I still support attempting to register the subscription as long as required minimal config keys are present, and that behavior is validated by [rhn_register](https://github.com/canonical/cloud-init/blob/main/cloudinit/config/cc_rh_subscription.py#L250-L253)

mostafaCamel replied:
Thanks for the reference. I should not have assumed that schema errors out on additional keys. I am ok with whatever route you recommend
